### PR TITLE
Feature/v1-193 | Implemented opportunity to get employee's available courses. Refactored search logic at AddCourseDialog for Employee at Employees page.

### PR DESCRIPTION
**Pull request by dapodev**
# Overview 

New endpoint implemented. Some utils added, some refactored.

# Changes

-Backend

1. New API endpoint `[GET] /api/employees/:id/courses/available?title=jq` where `:id` is employee's id and `title` is query parameter to filter courses by their title if necessary. *NO PAGINATION*
2. Compare `ObjectId` util refactored so to allow compare ids with passing one or both `string` values.
3. Implemented available courses filter util and normalizer.
4. Implemented new provider to get courses with statuses without pagination.

-Frontend

1. Refactored search logic at AddCourseDialog for Employee at Employees page.
2. Added api at frontend for getting available courses
3. Refactored types

### CODE REVIEW

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

File: backend/src/db/providers/courseProvider.ts [hunk 1]
```diff
@@ -210,9 +210,62 @@ const updateCourseField = async (courseId: string, field: string, value: unknown
   return updatedCourse;
 };
 
+const getAllCoursesProvider = async (userId: string, title?: string): Promise<ICourseStatus[]> => {
```
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/db/providers/courseProvider.ts:213]: No Pagination is here

File: backend/src/utils/normaliser/courses.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ICourseStatus, TAvailableCourse } from 'interfaces/ICourses/IQueryCourses';
+
+const filterOnlyAvailableCourses = (courses: ICourseStatus[]): ICourseStatus[] =>
```
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/utils/normaliser/courses.ts:3]: ICourseStatus is common course with `status` field

[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/utils/normaliser/courses.ts:3]: If status exists of course

File: backend/src/db/providers/courseProvider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/db/providers/courseProvider.ts:213]: but paggination in our getCourses is also odd

File: backend/src/utils/normaliser/courses.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/utils/normaliser/courses.ts:3]: rename to coursesWithStatus

File: backend/src/db/providers/courseProvider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/db/providers/courseProvider.ts:213]: It's not optional 😄

File: backend/src/utils/normaliser/courses.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/utils/normaliser/courses.ts]: done

File: backend/src/db/providers/courseProvider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/db/providers/courseProvider.ts]: image
 if "0" it return all courses

[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/db/providers/courseProvider.ts]: But it still requires pageSize 😄. And its not a good way to provide page size = 100000000000 or calculate it in advance

[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/db/providers/courseProvider.ts]: ok, what if default of limit will be 0? it should return all existed docs

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

File: backend/src/db/providers/courseProvider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/db/providers/courseProvider.ts]: doesnt work. https://screencast-o-matic.com/i/c3fYbfVY2cR

[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/db/providers/courseProvider.ts]: i believe that there are some solutions, for example just first one: 
image

[the remark this answers is not in the document]
Comment (SaXaPhonist) [reply backend/src/db/providers/courseProvider.ts]: you could move this populated fn out of provider

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

**Pull request closed without merging**